Migrate DQM to `edm::one::EDAnalyzer`

schneiml opened the pull request:
This is based on #22218 (thanks @fwyzard!), and supersedes #22157 (and also #22218, #22252).

- The migration will bring all of the memory savings of #22157. 
- It avoids the fallout of config changes (that is now isolated into #22243). Not that this is *not* required for this PR.
- A conversion to `edm::one::EDProducer<Accumulator>` should be a trivial change on top of #22243 and this one.
- The functional changes are 5d9538e, which performs the migration and removes now-obsolete logic, and bc0cd87, which fixes behaviour of per-lumi histograms and reintroduces some of the old logic for new purposes.

schneiml committed:
Remove thread_unsafe

fwyzard committed:
Migrate DQMEDAnalyzers from stream to one modules

Migrate all DQMEDAnalyzer modules from `edm::stream::EDAnalyzer<...>` to
`edm::one::EDAnalyzer<...>`.

This should significantly reduce the memory usage of a multithreaded DQM
job, at the price of preventing the framework from analysing multiple
runs or lumisections at the same time.

fwyzard committed:
Correctly handle per-lumi histograms.

This requires cloning the MEs of each plugin in endLuminosityBlock, for
which the moduleId is re-introduced.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22244/3394

cmsbuild commented:
A new Pull Request was created by  @schneiml (Marcel Schneider) for master.

It involves the following packages:

DQM/TrigXMonitor
DQMOffline/CalibCalo
DQMServices/Core


@vazzolini, @kmaeshima, @dmitrijus, @cmsbuild, @jfernan2, @vanbesien can you please review it and eventually sign? Thanks.
@mtosi, @barvic, @rociovilar this is something you requested to watch as well.
@davidlange6, @slava77, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22244/3395

Code check has found code style and quality issues which could be resolved by applying a patch in https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22244/3395/git-diff.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22244/3395/git-diff.patch | patch -p1`

You can run `scram build code-checks` to apply code checks directly

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22244/3397

Code check has found code style and quality issues which could be resolved by applying a patch in https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22244/3397/git-diff.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22244/3397/git-diff.patch | patch -p1`

You can run `scram build code-checks` to apply code checks directly

Dr15Jones commented:
> lumi-watching edm:: one::EDAnalyzer will block the framework from using them -- if I understand correctly, @Dr15Jones?

It does not block the framework from using concurrent LuminosityBlocks, it just makes them much less efficient. The framework can not call `beginLuminosityBlock` on a `one::` module for the next LuminosityBlock until the `endLuminosityBlock` of the first LuminosityBlock has been called. The framework is free to schedule other module's `beginLuminosityBlock` for the next LuminosityBlock while waiting for the `one`s to finish their `endLuminosityBlock` calls.

fwyzard commented:
Hi @Dr15Jones , ok, then it's less of a problem.
From the [documentation](https://twiki.cern.ch/twiki/bin/view/CMSPublic/FWMultithreadedFrameworkOneModuleInterface#edm_one_WatchLuminosityBlocks) of the `one` modules I was worried it would prevent the whole concurrent processing:

> **`edm::one::WatchLuminosityBlocks`**
This extension allows the module to watch LuminosityBlock transitions. If a module which extends `edm::one::WatchLuminosityBlocks` is added to a configuration, the system will only be allowed to process one LuminosityBlock at a time.

.A

Dr15Jones commented:
@fwyzard When I finally implemented the change, I was able to not be so restrictive.

fwyzard commented:
That's good news, thanks Chris.  
Indeed it makes me much less worried about the (possible future) impact of DQM `one` modules.  

On the other hand, I think it implies we cannot go with the simplified approach that Marcel proposed yesterday, for the lumisection-based `MonitorElements` ?

Based on Chris' description
> The framework can not call beginLuminosityBlock on a one:: module for the next LuminosityBlock until the endLuminosityBlock of the first LuminosityBlock has been called. The framework is free to schedule other module's beginLuminosityBlock for the next LuminosityBlock while waiting for the ones to finish their endLuminosityBlock calls.

is there anything that prevents this scenario ?
  - `one` modules **A** and **B** start analysing events from lumisection N
  - **A** completes lumisection N and starts analysing events from lumisection N+1, while **B** is still analysing (few) events from lumisection N

If that happens, we cannot have the `DQMStore` switch all lumisection-based MonitorElements from ls N to ls N+1 at the same time (this is what I was suggesting), nor have the various DQM output modules write all lumisection-based `MonitorElement`s at the same time and mark then as "lumisection N" (this is IIUC Marcel's suggestion).
Instead, we need to re-introduce the `endLuminosityBlock()` transition in the `DQMEDAnalyzer`, and make it clone the lumisection-based `MonitorElement`s owned by that module into a copy marked with the number of the lumisection that just finished.

Which in turn means we need to re-introduce the `moduleId` in the DQMStore element key, and a simplified version of the `mergeAndResetMEsLuminositySummaryCache()` method (possibly with a different name, it matters little).

I'll try to prototype that in #22218, if/when it works you can cherry-pick the commits here.

schneiml commented:
@fwyzard yes, I think we discussed exactly this scenario. The (potentially interesting) part is that the new `moduleId` serves a different purpose than the old one, so it could also be done differently. Though probably re-adding it is a fine solution.

makortel commented:
@Dr15Jones 

> When I finally implemented the change, I was able to not be so restrictive.

Could we update the documentation then?

cmsbuild commented:
The code-checks are being triggered in jenkins.

schneiml commented:
rebased to CMSSW_10_1_X_2018-02-15-2300 to clean up all this mess...

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22244/3409

cmsbuild commented:
Pull request #22244 was updated. @vazzolini, @kmaeshima, @dmitrijus, @cmsbuild, @jfernan2, @vanbesien can you please check and sign again.

fwyzard commented:
> Could we update the documentation then?

By the way, if a `one` module implements both, what is called first: `endLuminosityBlock()` or `endLuminosityBlockProduce()` ?

fwyzard commented:
@schneiml after looking at the modules' code, I don; tthink we can simply use the `endLuminosityBlock()` method to mark the luim-based histograms.

One possibility would be to mark it as `final`, and add a `virtual void dqmEndLuminosityBlock() { }` that the modules can override.

Or, we can use the `DQMStore`'s `postGlobalEndLumi()` signal.

schneiml commented:
@fwyzard we should get rid of modules overriding `endLuminosityBlock`. There are a number of them, but usually they don't do anything useful there. The only case where `DQMEDAnalyzer`s should care about lumisections is for per-lumi products.

fwyzard commented:
There are like 80 modules that overrode endLuminosityBlock; 4 of the 5 I checked actually do something in the method, so removing it does not sound like a good idea.

dmitrijus commented:
>  One possibility would be to mark it as final, and add a virtual void dqmEndLuminosityBlock() { } that the modules can override.

plus one

Dr15Jones commented:
@fwyzard 
>is there anything that prevents this scenario ?

> one modules **A** and **B** start analysing events from lumisection N
> A completes lumisection N and starts analysing events from lumisection N+1, while **B** is still analysing (few) events from lumisection N

I'm afraid the framework doesn't work the way you are thinking. However, the conclusion to draw is probably the same. 

All global `beginLuminosityBlock` of all modules which care about that transition must be run before the framework will run the stream `beginLuminosityBlock` calls and then start processing Events in that new LuminosityBlock. So **A** can not start analyzing events from LuminosityBlock N+1 until **B** has run its `beginLuminosityBlock` for N+1. However, **A** could be booking its new histograms in N+1 while **B** is still running its `endLuminosityBlock` for N.

Now if you had a one module **O** and a global module **G** then things are slightly different. **G** could start booking histograms to LuminosityBlock N+1 while **O** is still processing events from N. This is because global modules are allowed to see the next `beginLuminosityBlock` transition before the last `endLuminosityBlock` transition. Even in this scenario, **G** will not see any events from N+1 until **O** has run `beginLuminosityBlock` for N+1.

This behavior is outlined in this documentation:
https://twiki.cern.ch/twiki/bin/view/CMSPublic/FWMultithreadedFrameworkTransitions#Transitions

Dr15Jones commented:
@makortel 
> Could we update the documentation then?

Done. See
https://twiki.cern.ch/twiki/bin/view/CMSPublic/FWMultithreadedFrameworkModuleTypes#One_Modules

Dr15Jones commented:
@fwyzard 
> By the way, if a one module implements both, what is called first: `endLuminosityBlock()` or `endLuminosityBlockProduce()` ?

From
https://github.com/cms-sw/cmssw/blob/master/FWCore/Framework/src/one/EDProducerBase.cc#L133

`endLuminosityBlock()` is called then, `endLuminosityBlockProduce()`.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22244/3418

cmsbuild commented:
Pull request #22244 was updated. @vazzolini, @kmaeshima, @dmitrijus, @cmsbuild, @jfernan2, @vanbesien can you please check and sign again.

fwyzard commented:
> `endLuminosityBlock()` is called then, `endLuminosityBlockProduce()`.

Ah, good to know... so after migrating the `DQMEDAnalyzer` to `edm::one::EDProducer<edm::Accumulator>`, we could actually leave the `endLuminosityBlock()` as it was, and take care of the per-lumi histograms in the `endLuminosityBlockProduce()` method ?

I assume an `edm::one::EDAnaluyzer` does not have an `endLuminosityBlockProduce()` method ?

Dr15Jones commented:
@fwyzard 
> I assume an edm::one::EDAnaluyzer does not have an endLuminosityBlockProduce() method ?

Correct, an `EDAnalyzer` is not allowed to have a `produce` method for any transition.

fwyzard commented:
@Dr15Jones , how does this picture
![image](https://user-images.githubusercontent.com/4069793/36314364-e23db062-1334-11e8-99e7-a840e9057262.png)
look like for `one` module ?

dmitrijus commented on DQMServices/Core/interface/DQMEDAnalyzer.h:
```diff
@@ -1,80 +1,54 @@
+  void endRun(edm::Run const& run, edm::EventSetup const& setup) override
+  { }
+
+  void beginLuminosityBlock(edm::LuminosityBlock const& lumi, edm::EventSetup const& setup) override
```
can this be final?

fwyzard replied:
Feel free to make a PR.

schneiml replied:
At least one module overrides this with non-trivial logic, so no, not for now.

dmitrijus commented on DQMServices/Core/src/DQMStore.cc:
```diff
@@ -2182,6 +2019,47 @@ DQMStore::postGlobalBeginLumi(const edm::GlobalContext &gc)
+    data_.insert(std::move(clone));
+
+    // reset the ME for the next lumisection
+    const_cast<MonitorElement*>(&*i)->Reset();
```
this is great, actually.

Can we remove DQMStore::postGlobalBeginLumi, https://github.com/cms-sw/cmssw/blob/master/DQMServices/Core/src/DQMStore.cc#L2159

Since it is now done here?

fwyzard replied:
Feel free to make a PR.

However, either that was already unnecessary (because of the call to `Reset()` in the `mergeAndResetMEsLuminositySummaryCache()` method) or is still needed (possibly for the legacy modules)

Dr15Jones commented:
@fwyzard 
> how does this picture ... look like for one module ?

If the one module wants the LuminosityBlock transitions, then the global Blue begin LuminosityBlock box still has the same left hand side (i.e. starts at the same time) but the right hand side now stretches to overlap the green global end LuminosityBlock (i.e. it can not continue until the one module's end LuminosityBlock is called at which point the blue LuminosityBlock can do the begin call for that module).

fwyzard commented:
OK, I guess I'm confused :-(

Hopefully that will not lean to any broken code...

cmsbuild commented:
The code-checks are being triggered in jenkins.

fwyzard commented:
From the commit history, I would guess it needs some kind of rebase ?

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22244/3423

cmsbuild commented:
Pull request #22244 was updated. @ghellwig, @smuzaffar, @civanch, @Dr15Jones, @vazzolini, @lveldere, @kmaeshima, @arunhep, @franzoni, @mdhildreth, @dmitrijus, @nsmith-, @rekovic, @cmsbuild, @jfernan2, @cerminar, @thomreis, @vanbesien, @monttj, @ssekmen, @lpernie can you please check and sign again.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22244/3442

Code check has found code style and quality issues which could be resolved by applying a patch in https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22244/3442/git-diff.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22244/3442/git-diff.patch | patch -p1`

You can run `scram build code-checks` to apply code checks directly

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22244/3443

Code check has found code style and quality issues which could be resolved by applying a patch in https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22244/3443/git-diff.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22244/3443/git-diff.patch | patch -p1`

You can run `scram build code-checks` to apply code checks directly

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22244/3444

cmsbuild commented:
Pull request #22244 was updated. @cmsbuild, @civanch, @vazzolini, @kmaeshima, @mdhildreth, @dmitrijus, @nsmith-, @rekovic, @jfernan2, @kpedro88, @thomreis, @vanbesien can you please check and sign again.

kpedro88 commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: dafb694ec4bb45981cd5a3c88689d8daecb8b2e2
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22244/26169/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/b35d5b8dc91f50a7fb6b7e3e3a54926e22df2f0b
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22244/26169/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22244/26169/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22244/26169/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 32 differences found in the comparisons
 - DQMHistoTests: Total files compared: 27
 - DQMHistoTests: Total histograms compared: 2465360
 - DQMHistoTests: Total failures: 9
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2465182
 - DQMHistoTests: Total skipped: 169
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.509999999726 KiB( 22 files compared)
 - Checked 111 log files, 9 edm output root files, 27 DQM output files

civanch commented:
+1

davidlange6 commented:
Your PR is unmergeable. Please have a look and possibly rebase it.

fwyzard commented:
I *explicitly* do not give permission to use commit f7670388c2aadf84729f48983b8424296b987e47 .
I disagree with its intent, and there are better alternatives.

fwyzard commented:
hold

cmsbuild commented:
Pull request has been put on hold by  @fwyzard
They need to issue an `unhold` command to remove the `hold` state or L1 can `unhold` it for all

schneiml commented:
re commit f767038: as far as I see, the advantages are:
- it works, right now. The alternative is rather obvious but not tested.
- It is consistent with harvesting (https://github.com/cms-sw/cmssw/blob/master/DQMServices/Core/interface/DQMEDHarvester.h#L32), where end lumi transitions are more commonly used.
- It is the obvious way to achieve the required behaviour: make sure the original handler cannot be overrriden to ensure that the original code runs, then delegate to a different method.

Disadvantages are:
- More invasive changes to subsystem code. Though the alternative needs #22243 first, which is rather huge as well.
- Another name for the same thing for people to remember.
- more?

-

fwyzard commented:
The desired behaviour is to allow users to define actions to be run at the end of the lumisection.
The obvious way to achieve it is simply to leave the `endLuminosityBlock()` method available.

From a user's point of view, it makes no sense to have to implement a pair of methods called `beginLuminosityBlock()` and `dqmEndLuminosityBlock()`.
If you think that all DQM modules should use methods prepended with `dqm`, then please go ahead and (re)define `dqmBeinRun()`, `dqmEndRun()`, `dqmBeginLuminosityBlock()`, `dqmEndLuminosityBlock()`, and *document* them.

schneiml commented:
@fwyzard Yes, keeping the original handler is more clear from the user's point of view, but less obvious in terms of implementation. And I do want to keep the implementation as simple and obvious as possible, given that documentation is in any case either missing, outdated, or impossible to find.

In my opinion, all handlers should  be prefixed, again consistent with harvesting, where they are more commonly used. I can prepare these changes, if you are happy with it.

fwyzard commented:
If you want to make these global changes to all DQM modules I don't have strong objections.

However I would not couple this change with the migration to `one` and to `EDProducer`.

schneiml commented:
@fwyzard Wait, now I don't get your last sentence: We want to get the `edm::one` migration (that is why we started this mess of PRs), and now we have to tie it with the renaming *or* the `EDProducer` migration, for technical reasons. 

It seems you prefer using the `EDProducer` solution, but stopped just short of testing it. I prefer the renaming, since it is easier to predict the implications for that.

fwyzard commented:
@schneiml , here is what I would do...

## if you want to rename `methods` to `dqmMethods`
- make a PR with the rename, before the migration from `stream` to `one`
- make a PR with the migration from `stream` to `one`
  - call `DQMStore::cloneLumiHistograms()` inside `endLuminosityblock()` after calling `dqmEndLuminosityBlock()`
- eventually, make a PR with the migration from `EDAnalyzer` to `EDProducer<Accumulator>`
  - move the call to `DQMStore::cloneLumiHistograms()` to `endLuminosityBlockProduce()`, before producing the DQM token for the lumi histograms

## if you do not want to rename `methods` to `dqmMethods`
- start with #22218, and add the call to `DQMStore::cloneLumiHistograms()` inside `DQMStore::postModuleGlobalEndLumi()`
- eventually, make a PR with the migration from `EDAnalyzer` to `EDProducer<Accumulator>`
  - optionally move the call to `DQMStore::cloneLumiHistograms()` back to `endLuminosityBlockProduce()`, before producing the DQM token for the lumi histograms

schneiml commented:
@fwyzard ah, so you *do* have a third solution in mind: using `DQMStore::postModuleGlobalEndLumi()`. Now, I am rather ingorant of what the semantics of this hook are, but didn't we come to the conclusion that this is *not correct*, since the framework might use concurrent lumis whenever it likes?

fwyzard commented:
Yes, as I wrote in the description of #22218.

The call to a service's `postModuleGlobalEndLumi` method for a given module happens right after that module's (global) `endLuminosityBlock` method, see https://github.com/cms-sw/cmssw/blob/master/FWCore/Framework/src/Worker.h#L816_L817 :
  - `cpp`'s constructor calls each services' `preModuleGlobalEndLumi` method
  - `implDoEnd(...)` calls the module's global `endLuminosityBlock` method
  - `cpp`'s destructor calls each services' `preModuleGlobalEndLumi` method

schneiml commented:
@fwyzard ah, this sounds almost like it is *designed* for whatever magic we do here. So I'd prefer going that route. The renaming (as well as the `EDProducer` migration can then happen independently (or not at all).

fwyzard commented:
OK.

Do you want me to code the changes to the DQM service ?

schneiml commented:
@fwyzard if you know what to do, just add a commit with it to 22218.

schneiml commented:
@thomreis for the record, we have an interesting curiosity here: https://github.com/cms-sw/cmssw/blob/master/DQMOffline/L1Trigger/interface/L1TMuonDQMOffline.h#L76
(related to the renaming business, so far not a problem, but that might not work as expected...)

thomreis commented:
Hi @schneiml,
you mean that the name is already used in the L1 offline muon DQM code?
It looks as of this does not do anything particularly essential anyway https://github.com/cms-sw/cmssw/blob/master/DQMOffline/L1Trigger/src/L1TMuonDQMOffline.cc#L238. If problems arise we might as well remove that function.

fwyzard committed:
Handle lumisection-based histograms in the DQMStore

For the time being, lumisection-based histograms are cloned and set with
the correct lumisection index in the `postModuleGlobalEndLumi` signal
handler of the `DQMStore`.

If/when the `DQMEDAnalyzer` modules are migrated from `EDAnalyzer` to
`EDProducer<Accumulator>`, the call should be moved to the modules'
`endLuminosityBlockProduce` method.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22244/3471

cmsbuild commented:
Pull request #22244 was updated. @vazzolini, @kmaeshima, @dmitrijus, @cmsbuild, @jfernan2, @vanbesien can you please check and sign again.

schneiml commented:
PR updated with "the third way", as outlined by @fwyzard (Thanks again!) in https://github.com/cms-sw/cmssw/pull/22244#issuecomment-366935000.

This means much less impact on subsystem code, which is great. The solution used now is not that elegant (more work for the DQMService, where we'd rather like to have it do less), but it should not have any technical limitations (even correct for concurrent lumisections).

This resolves all the known problmems (including the merge conflict).

schneiml commented:
@thomreis I suspect that is an artifact of somebody copying the harvesting interface to step one. It can't do much, since it likely never runs... just remove it on a good occasion.

dmitrijus commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

dmitrijus commented:
+1

fwyzard commented:
since most of the commits are mine, I would prefer to have #22218 merged...

cmsbuild commented:
+1
Tested at: c67f435b7bae9c4e7b07d25e2d2e5e9f6663b4cf
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22244/26194/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @davidlange6, @slava77, @smuzaffar, @fabiocos (and backports should be raised in the release meeting by the corresponding L2)

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22244/26194/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 32 differences found in the comparisons
 - DQMHistoTests: Total files compared: 28
 - DQMHistoTests: Total histograms compared: 2498032
 - DQMHistoTests: Total failures: 9
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2497847
 - DQMHistoTests: Total skipped: 176
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.829999999929 KiB( 23 files compared)
 - Checked 115 log files, 9 edm output root files, 28 DQM output files
